Migration of instance to other cluster members WD-4264

edlerd opened the pull request:
## Done

- show location in configuration section of instance detail page when clustered environment
- add migrate button next to it
- handle instance migration

Fixes WD-4264

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @lorumic or @edlerd for access.
    - With a local copy of this branch, run as described in the [Readme](https://github.com/canonical/lxd-ui#setting-up-for-development).
2. Perform the following QA steps:
    - test in clustered and unclustered environments
    - migrate instance to some cluster member

webteam-app commented:
Demo starting at https://lxd-ui-370.demos.haus

edlerd commented:
See screens below for design review

![Screenshot from 2023-06-06 18-42-19](https://github.com/canonical/lxd-ui/assets/1155472/15c0168a-b56d-43bd-a717-0bd5b2ba4472)
![Screenshot from 2023-06-06 18-41-47](https://github.com/canonical/lxd-ui/assets/1155472/ffed7e69-849e-462c-b3a2-e997b0af9c22)
![Screenshot from 2023-06-06 18-41-41](https://github.com/canonical/lxd-ui/assets/1155472/92d035d0-4471-48a3-be66-c4c4f9a2fb20)
![Screenshot from 2023-06-06 18-41-35](https://github.com/canonical/lxd-ui/assets/1155472/67aa8abe-27a0-4204-84df-66df42ec45bb)
![Screenshot from 2023-06-06 18-41-30](https://github.com/canonical/lxd-ui/assets/1155472/6a1c7f0e-e4e8-4775-8c07-d7bdb92ca220)

al-bakalova commented:
@edlerd That's the experience that we re looking for. However, the layout seems a bit off:

- The Migration notifications must be within the tab and not above the tabs as the action belongs to the tab and not to the header details
- I believe, when migrating an instance, we need to chose both the group and the member. Maybe confirm with the Core team
- I would prefer for us to have a slightly bigger migrate modal. It looks too condensed, let's let it breathe a little. I would double the width. That way, we can accommodate for longer instance names
- In the header, why are the status and the actions going to a next line. Looks like there is more than enough space for us to display all the details, the breadcrumb and the actions, on a single row.
- When not in "Edit mode", the Migrate button should be hidden.

edlerd commented:
@al-bakalova some updates addressing most of the points.

Re "group and the member", that is an open bug on lxd core, currently only member migration is supported.

![Screenshot from 2023-06-06 22-18-19](https://github.com/canonical/lxd-ui/assets/1155472/fee40cd7-dd73-493b-ac4c-6b74a2eb92ce)
![Screenshot from 2023-06-06 22-17-30](https://github.com/canonical/lxd-ui/assets/1155472/5794e972-2fd5-481b-9e9c-6dc375392407)
![Screenshot from 2023-06-06 22-17-17](https://github.com/canonical/lxd-ui/assets/1155472/7238bb92-cca5-470a-8cfb-8b22180f4da1)

al-bakalova commented:
@edlerd Thanks for applying all the changes. I think all the shared screens are good to go.

Were you able to slightly update the width of the Migrate modal?

edlerd commented:
Yes, gave it a min width - below a screenshot of the recent one.

![Screenshot from 2023-06-07 09-36-58](https://github.com/canonical/lxd-ui/assets/1155472/045fae2a-f8db-45ed-8ac3-84a99bb80f51)

al-bakalova commented:
@edlerd Perfect, thank you! You are getting a design +1 :)

lorumic commented on src/pages/instances/actions/MigrateInstanceBtn.tsx:
```diff
@@ -0,0 +1,90 @@
+
+  const handleSuccess = () => {
+    setInTabNotification(
+      success(`Migration finished for instance ${instance}`)
```
I think we should use the `ItemName` component, with the `bold` prop, here and in all the other notifications where we mention the instance name.

edlerd replied:
In the failures, the name is part of the title which is bold itself, but here we can use it, sure.

lorumic commented on src/pages/instances/MigrateInstanceForm.tsx:
```diff
@@ -0,0 +1,81 @@
+          <Button
+            className="u-no-margin--bottom"
+            type="button"
+            aria-label="cancel reconnect"
```
```suggestion
            aria-label="cancel migrate"
```

edlerd commented on src/pages/instances/actions/MigrateInstanceBtn.tsx:
```diff
@@ -0,0 +1,90 @@
+        closePortal();
+      })
+      .catch((e) => {
+        setInTabNotification(
+          failure(`Migration failed on instance ${instance}`, e)
+        );
```
They seem, but are not directly compatible. The `handleFailure` uses a slightly different interface. It expects a message argument as a string and turns it into an error. While here, we pass the exception object directly to the failure method.

lorumic replied:
But you can make the function compatible to both cases, right?

edlerd commented:
> Do you know why this is happening? Could it be that the corresponding formik value is not being updated after a query invalidation?

Yes, formik was still having a shadow copy of the instance with the old location. The way you asked here pointed at a easy fix to update that value after the migration finished.
